# 8.2 Exercise: Best Model Selection and Hyperparameter Tuning
 

# Importing the dataset 

In [88]:
# Load the dataset as a Pandas data frame.

import pandas as pd

loan = pd.read_csv(r"C:\Users\Riaz\Desktop\MSDS\Data Mining\Week 8\Loan_Train.csv")

# Required libraries are imported and using the read_csv function the dataframe has been loaded from csv file.

print ('The shape of the original dataset is',loan.shape)

loan

The shape of the original dataset is (614, 13)


Loan_ID  Gender Married Dependents     Education Self_Employed  \
0    LP001002    Male      No          0      Graduate            No   
1    LP001003    Male     Yes          1      Graduate            No   
2    LP001005    Male     Yes          0      Graduate           Yes   
3    LP001006    Male     Yes          0  Not Graduate            No   
4    LP001008    Male      No          0      Graduate            No   
..        ...     ...     ...        ...           ...           ...   
609  LP002978  Female      No          0      Graduate            No   
610  LP002979    Male     Yes         3+      Graduate            No   
611  LP002983    Male     Yes          1      Graduate            No   
612  LP002984    Male     Yes          2      Graduate            No   
613  LP002990  Female      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5849                0.0         NaN             360.0   
1               4583             1508.0       128.0             360.0   
2               3000                0.0        66.0             360.0   
3               2583             2358.0       120.0             360.0   
4               6000                0.0       141.0             360.0   
..               ...                ...         ...               ...   
609             2900                0.0        71.0             360.0   
610             4106                0.0        40.0             180.0   
611             8072              240.0       253.0             360.0   
612             7583                0.0       187.0             360.0   
613             4583                0.0       133.0             360.0   

     Credit_History Property_Area Loan_Status  
0               1.0         Urban           Y  
1               1.0         Rural           N  
2               1.0         Urban           Y  
3               1.0         Urban           Y  
4               1.0         Urban           Y  
..              ...           ...         ...  
609             1.0         Rural           Y  
610             1.0         Rural           Y  
611             1.0         Urban           Y  
612             1.0         Urban           Y  
613             0.0     Semiurban           N  

[614 rows x 13 columns]

# Preparing the data for modeling 

In [89]:
# Dropping the column “Load_ID.” using drop function

loan_drop = loan.drop('Loan_ID',axis=1)

# Finding if there are any missing data in the dataset,

print ("The missing data in the dataset is \n\n",loan_drop.isna().sum(),'\n\n')

# Dropping the rows with missing data.

loan_dropped = loan_drop.dropna()

print ("Dataset after dropping NaN rows")

loan_dropped


The missing data in the dataset is 

 Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64 


Dataset after dropping NaN rows


Gender Married Dependents     Education Self_Employed  ApplicantIncome  \
1      Male     Yes          1      Graduate            No             4583   
2      Male     Yes          0      Graduate           Yes             3000   
3      Male     Yes          0  Not Graduate            No             2583   
4      Male      No          0      Graduate            No             6000   
5      Male     Yes          2      Graduate           Yes             5417   
..      ...     ...        ...           ...           ...              ...   
609  Female      No          0      Graduate            No             2900   
610    Male     Yes         3+      Graduate            No             4106   
611    Male     Yes          1      Graduate            No             8072   
612    Male     Yes          2      Graduate            No             7583   
613  Female      No          0      Graduate           Yes             4583   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
1               1508.0       128.0             360.0             1.0   
2                  0.0        66.0             360.0             1.0   
3               2358.0       120.0             360.0             1.0   
4                  0.0       141.0             360.0             1.0   
5               4196.0       267.0             360.0             1.0   
..                 ...         ...               ...             ...   
609                0.0        71.0             360.0             1.0   
610                0.0        40.0             180.0             1.0   
611              240.0       253.0             360.0             1.0   
612                0.0       187.0             360.0             1.0   
613                0.0       133.0             360.0             0.0   

    Property_Area Loan_Status  
1           Rural           N  
2           Urban           Y  
3           Urban           Y  
4           Urban           Y  
5           Urban           Y  
..            ...         ...  
609         Rural           Y  
610         Rural           Y  
611         Urban           Y  
612         Urban           Y  
613     Semiurban           N  

[480 rows x 12 columns]

In [90]:
# Converting the categorical features into dummy variables

# Converting the Dependents column to categorical values, as it is ordinal.

print ("Number of unique values present in Dependents column\n\n",loan_dropped['Dependents'].value_counts())

loan_dropped['Dependents'] = loan_dropped['Dependents'].replace({'3+':3})




Number of unique values present in Dependents column

 Dependents
0     274
2      85
1      80
3+     41
Name: count, dtype: int64


C:\Users\Riaz\AppData\Local\Temp\ipykernel_11180\3112777725.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  loan_dropped['Dependents'] = loan_dropped['Dependents'].replace({'3+':3})


In [91]:
# Converting the categorical features into dummy variables using get_dummies method,

loan_dropped_dummies = pd.get_dummies(loan_dropped, columns=['Gender','Married','Dependents','Education','Self_Employed','Property_Area','Credit_History'],prefix=['Gender','Married','Dependents','Education','Self_Employed','Property_Area','Credit_History'], drop_first=True)
loan_dropped_dummies

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1               4583             1508.0       128.0             360.0   
2               3000                0.0        66.0             360.0   
3               2583             2358.0       120.0             360.0   
4               6000                0.0       141.0             360.0   
5               5417             4196.0       267.0             360.0   
..               ...                ...         ...               ...   
609             2900                0.0        71.0             360.0   
610             4106                0.0        40.0             180.0   
611             8072              240.0       253.0             360.0   
612             7583                0.0       187.0             360.0   
613             4583                0.0       133.0             360.0   

    Loan_Status  Gender_Male  Married_Yes  Dependents_0  Dependents_1  \
1             N         True         True         False          True   
2             Y         True         True          True         False   
3             Y         True         True          True         False   
4             Y         True        False          True         False   
5             Y         True         True         False         False   
..          ...          ...          ...           ...           ...   
609           Y        False        False          True         False   
610           Y         True         True         False         False   
611           Y         True         True         False          True   
612           Y         True         True         False         False   
613           N        False        False          True         False   

     Dependents_2  Education_Not Graduate  Self_Employed_Yes  \
1           False                   False              False   
2           False                   False               True   
3           False                    True              False   
4           False                   False              False   
5            True                   False               True   
..            ...                     ...                ...   
609         False                   False              False   
610         False                   False              False   
611         False                   False              False   
612          True                   False              False   
613         False                   False               True   

     Property_Area_Semiurban  Property_Area_Urban  Credit_History_1.0  
1                      False                False                True  
2                      False                 True                True  
3                      False                 True                True  
4                      False                 True                True  
5                      False                 True                True  
..                       ...                  ...                 ...  
609                    False                False                True  
610                    False                False                True  
611                    False                 True                True  
612                    False                 True                True  
613                     True                False               False  

[480 rows x 15 columns]

In [92]:
# Randomly split the data into 80% training data and 20% test data

from sklearn.model_selection import train_test_split

# Using the train_test_split function we are passing the dataframe and the target which is the mpg and mentioning the 
# test size as 0.2 meaning 20%.

loan_dropped_dummies_train, loan_dropped_dummies_test, Loan_Status_train, Loan_Status_test = train_test_split(loan_dropped_dummies.drop(columns=['Loan_Status']), loan_dropped_dummies['Loan_Status'], test_size=0.2, random_state=42)




# Creating a pipeline with a min-max scaler and a KNN classifier 

In [93]:
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import MinMaxScaler

# Create a KNN classifier
# Using the default KNN classifier of n = 5

knn = KNeighborsClassifier(n_neighbors=5)

# Create a pipeline
pipe = Pipeline([("MinMaxScaler", MinMaxScaler()), ("knn", knn)])

# Fit the pipeline on the training data
pipe.fit(loan_dropped_dummies_train, Loan_Status_train)

Pipeline(steps=[('MinMaxScaler', MinMaxScaler()),
                ('knn', KNeighborsClassifier())])

In [94]:
# Predicting using the pipe we have created

Loan_Status_test_predicted=pipe.predict(loan_dropped_dummies_test)

#Calculating accuracy score

from sklearn.metrics import accuracy_score
accuracy = accuracy_score(Loan_Status_test, Loan_Status_test_predicted)
print(f'Accuracy of default KNN model: {accuracy}')

Accuracy of default KNN model: 0.7708333333333334


# Creating a search space for KNN classifier 

In [95]:
# Creating a search space for KNN classifier from 1 to 10
from sklearn.model_selection import GridSearchCV

# Create space of candidate values
search_space = [{"knn__n_neighbors": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}]

knn1 = KNeighborsClassifier()
pipe1 = Pipeline([("MinMaxScaler", MinMaxScaler()), ("knn", knn1)])

# Create grid search for the training dataset
classifier = GridSearchCV(
    pipe1, search_space, cv=5, verbose=0).fit(loan_dropped_dummies_train, Loan_Status_train)

# Best neighborhood size (k)
print ("The best value of n neighbours as per the GridSearchCV is\n",classifier.best_estimator_.get_params()["knn__n_neighbors"])

The best value of n neighbours as per the GridSearchCV is
 10


As the GridSearchCV has given the best value of n as 10, we are trying with that value

In [96]:
# Fitting the GridSearchCV to training dataset

classifier.fit(loan_dropped_dummies_train, Loan_Status_train)

print ("The parameters of the GridSearchCV are \n\n",classifier.get_params())

The parameters of the GridSearchCV are 

 {'cv': 5, 'error_score': nan, 'estimator__memory': None, 'estimator__steps': [('MinMaxScaler', MinMaxScaler()), ('knn', KNeighborsClassifier())], 'estimator__verbose': False, 'estimator__MinMaxScaler': MinMaxScaler(), 'estimator__knn': KNeighborsClassifier(), 'estimator__MinMaxScaler__clip': False, 'estimator__MinMaxScaler__copy': True, 'estimator__MinMaxScaler__feature_range': (0, 1), 'estimator__knn__algorithm': 'auto', 'estimator__knn__leaf_size': 30, 'estimator__knn__metric': 'minkowski', 'estimator__knn__metric_params': None, 'estimator__knn__n_jobs': None, 'estimator__knn__n_neighbors': 5, 'estimator__knn__p': 2, 'estimator__knn__weights': 'uniform', 'estimator': Pipeline(steps=[('MinMaxScaler', MinMaxScaler()),
                ('knn', KNeighborsClassifier())]), 'n_jobs': None, 'param_grid': [{'knn__n_neighbors': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}], 'pre_dispatch': '2*n_jobs', 'refit': True, 'return_train_score': False, 'scoring': None, 've

In [97]:
# Predicting using the GridSearchCV  we have created,

Loan_Status_test_predicted_grid=classifier.predict(loan_dropped_dummies_test)

#Calculating accuracy score of GridSearchCV

accuracy_grid = accuracy_score(Loan_Status_test, Loan_Status_test_predicted_grid)
print(f'Accuracy of default GridSearchCV model, but with the default n value as 5 as seen from the above output: {accuracy}')

Accuracy of default GridSearchCV model, but with the default n value as 5 as seen from the above output: 0.7708333333333334


In [98]:
# Using the n value = 10, as suggested by the GridSearchCV and getting the accuracy,

knn2 = KNeighborsClassifier(n_neighbors=10)
pipe2 = Pipeline([("MinMaxScaler", MinMaxScaler()), ("knn", knn2)])

# Create grid search for the training dataset
classifier1 = GridSearchCV(
    pipe2, search_space, cv=5, verbose=0).fit(loan_dropped_dummies_train, Loan_Status_train)

classifier1.fit(loan_dropped_dummies_train, Loan_Status_train)
print ("The parameters of the GridSearchCV are \n\n",classifier1.get_params(),'\n\n')

# Predicting using the GridSearchCV  we have created,

Loan_Status_test_predicted_grid1=classifier1.predict(loan_dropped_dummies_test)

#Calculating accuracy score of GridSearchCV

accuracy_grid = accuracy_score(Loan_Status_test, Loan_Status_test_predicted_grid1)
print(f'Accuracy of default GridSearchCV model, with the n value as 10: {accuracy}')


The parameters of the GridSearchCV are 

 {'cv': 5, 'error_score': nan, 'estimator__memory': None, 'estimator__steps': [('MinMaxScaler', MinMaxScaler()), ('knn', KNeighborsClassifier(n_neighbors=10))], 'estimator__verbose': False, 'estimator__MinMaxScaler': MinMaxScaler(), 'estimator__knn': KNeighborsClassifier(n_neighbors=10), 'estimator__MinMaxScaler__clip': False, 'estimator__MinMaxScaler__copy': True, 'estimator__MinMaxScaler__feature_range': (0, 1), 'estimator__knn__algorithm': 'auto', 'estimator__knn__leaf_size': 30, 'estimator__knn__metric': 'minkowski', 'estimator__knn__metric_params': None, 'estimator__knn__n_jobs': None, 'estimator__knn__n_neighbors': 10, 'estimator__knn__p': 2, 'estimator__knn__weights': 'uniform', 'estimator': Pipeline(steps=[('MinMaxScaler', MinMaxScaler()),
                ('knn', KNeighborsClassifier(n_neighbors=10))]), 'n_jobs': None, 'param_grid': [{'knn__n_neighbors': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}], 'pre_dispatch': '2*n_jobs', 'refit': True, 'retur

As seen from the above model, accuracy is same when using the n value as 5 or 10

# Expanding search space to include logistic regression and random forest models

In [99]:
# Repeating steps 6 and 7 with the same pipeline, 
# but including logistic regression and random forest models with the hyperparameter values in search space.

from sklearn.linear_model import LogisticRegression
import numpy as np
from sklearn.ensemble import RandomForestClassifier

pipe3 = Pipeline([("MinMaxScaler", MinMaxScaler()), ("classifier", knn2)])

search_space_models = [{"classifier": [LogisticRegression(max_iter=500,
       solver='liblinear')],
                 "classifier__penalty": ['l1', 'l2'],
                 "classifier__C": np.logspace(0, 4, 10)},
                {"classifier": [RandomForestClassifier()],
                 "classifier__n_estimators": [10, 100, 1000],
                 "classifier__max_features": [1, 2, 3]},
                      {"classifier": [KNeighborsClassifier()],
                 "classifier__n_neighbors": [10]}]

classifier2 = GridSearchCV(
    pipe3, search_space_models, cv=5, verbose=0).fit(loan_dropped_dummies_train, Loan_Status_train)


In [100]:
print ("The detailed results \n", classifier2.cv_results_)

The detailed results 
 {'mean_fit_time': array([0.00845294, 0.00463562, 0.00840874, 0.00400491, 0.00821967,
       0.00497994, 0.0058537 , 0.00748153, 0.01075573, 0.00591702,
       0.00655494, 0.00398068, 0.00597472, 0.00432367, 0.00639057,
       0.00397453, 0.0060112 , 0.00380712, 0.00634742, 0.00556593,
       0.02971816, 0.1715889 , 1.67139878, 0.02760792, 0.16350389,
       1.66854029, 0.0183321 , 0.1846736 , 1.78533082, 0.00400343]), 'std_fit_time': array([2.87710847e-03, 1.29816533e-03, 1.60366531e-03, 7.44005643e-05,
       2.30135334e-03, 1.26243945e-03, 8.69067992e-04, 1.61565423e-03,
       1.31924782e-03, 2.16530679e-03, 6.52082121e-04, 3.34602447e-06,
       1.14305734e-05, 4.76390472e-04, 4.72470459e-04, 1.98381451e-05,
       3.68716934e-05, 4.05554626e-04, 4.81427958e-04, 1.96515638e-03,
       1.32240626e-02, 1.27198313e-02, 1.42643952e-01, 5.97151139e-03,
       1.71404259e-02, 5.27586574e-02, 1.00436773e-03, 2.34932226e-02,
       4.29266035e-02, 6.30655037e-04]), '

In [101]:
# Print best model
print("The best estimator\n",classifier2.best_estimator_,'\n')
print("The best score\n",classifier2.best_score_,'\n')
print("The best params\n",classifier2.best_params_,'\n')



The best estimator
 Pipeline(steps=[('MinMaxScaler', MinMaxScaler()),
                ('classifier',
                 LogisticRegression(C=7.742636826811269, max_iter=500,
                                    penalty='l1', solver='liblinear'))]) 

The best score
 0.8100136705399864 

The best params
 {'classifier': LogisticRegression(C=7.742636826811269, max_iter=500, penalty='l1',
                   solver='liblinear'), 'classifier__C': 7.742636826811269, 'classifier__penalty': 'l1'} 



In [102]:
# Accuracy of the above GridCVSearch selected model on the test set

testmodel = classifier2.best_estimator_
Loan_Status_test_predicted_grid1_test=testmodel.predict(loan_dropped_dummies_test)

#Calculating accuracy score of GridSearchCV

accuracy_grid1 = accuracy_score(Loan_Status_test, Loan_Status_test_predicted_grid1_test)
print(f'Accuracy of best model derived from GridSearchCV with the searchspace of different algorithms and different hyperparameters: {accuracy_grid1}')

Accuracy of best model derived from GridSearchCV with the searchspace of different algorithms and different hyperparameters: 0.8229166666666666


# Analysis

Imported the dataset and prepared for modelling by roping the column “Load_ID.” and NaN values.  Converted the categorical features into dummy variables.  
Created a pipeline with a min-max scaler and a KNN classifier.  The accuracy of default KNN values were 0.770.  
Created a search space for KNN classifier with “n_neighbors” parameter varies from 1 to 10.   The best value of n neighbors was 10 and the predicted accuracy of KNN with this value was 0.770, when CV of 5 fold was used.  
Expanded search space to include logistic regression and random forest models alongwith the hyperparameter values.  The best parameters were found to be,  

 {'classifier': LogisticRegression(C=7.742636826811269, max_iter=500, penalty='l1',  
                    solver='liblinear'), 'classifier__C': 7.742636826811269, 'classifier__penalty': 'l1'}   
                    
The accuracy improved to 0.822 with this above model.  
